src/db/mysql/models.rs

```diff
@@ -327,43 +326,31 @@ impl MysqlDb {
         let user_id: u64 = bso.user_id.legacy_id;
         let timestamp = self.timestamp().as_i64();
 
-        // XXX: consider mysql ON DUPLICATE KEY UPDATE?
         self.conn.transaction(|| {
+            let payload = bso.payload.as_ref().map(Deref::deref).unwrap_or_default();
+            let sortindex = bso.sortindex;
+            let ttl = bso.ttl.map_or(DEFAULT_BSO_TTL, |ttl| ttl);
             let q = r#"
-                SELECT 1 as count FROM bso
-                WHERE user_id = ? AND collection_id = ? AND id = ?
+                INSERT INTO bso (user_id, collection_id, id, sortindex, payload, modified, expiry)
+                VALUES (?, ?, ?, ?, ?, ?, ?)
+                ON DUPLICATE KEY UPDATE
```

Comment:
A couple issues here:

- ON DUPLICATE KEY UPDATE targets the existing row (by primary key) for an UPDATE if it already exists. So we want to indicate an UPDATE of all the bso fields in the `PutBso`/`UpdateBso` structs *except* the primary key (user_id/collection_id/id, which are already implied): so they shouldn't be specified here. This would look more like `ON DUPLICATE KEY UPDATE sortindex = VALUES(sortindex) payload = ..etc`

You can think of the UPDATE clause being generated as `UPDATE bso set sortindex = ? payload = ..etc WHERE user_id = ? AND collection_id = ..etc`

- Additionally all 4 of the fields in question for update are *optional* -- they're only included in the update clause if they're not a `None` value in PutBso. So we need to generate the update part of the clause dynamically. With the `VALUES()` feature of ON DUPLICATE KEY UPDATE we shouldn't need to dynamically bind the update values (which probably requires a boxed query)

Reply:
Thanks for the explanation!